### PR TITLE
Tray icon

### PR DESCRIPTION
**Pull request by tstumm**
This pull request implements a hide-to-tray functionality. This feature is configurable in the settings and defaults to being not activated.

The tray icon is only shown if the window is minimized, which also hides the window from the taskbar.
Some users may want to keep the tray icon visible all the time or to prevent closing the window in favour of showing the icon. Both improvements can be done very quickly.

I also included the localization strings for English (please be aware that I'm not a native speaker).

### CODE REVIEW

File: src/app/accounts/settings.component.ts [hunk 1]
```diff
@@ -83,6 +86,11 @@ export class SettingsComponent implements OnInit {
         this.callAnalytics('Favicons', !this.disableGa);
     }
 
+    async saveHideInTray() {
+        await this.storageService.save(DesktopConstantsService.enableHideInTrayKey, this.enableHideInTray);
+        this.callAnalytics('HideInTray', !this.disableGa);
```
[the remark this answers is not in the document]
Comment (tstumm) [reply src/app/accounts/settings.component.ts]: Jup, you are right. I copied this from saveFavicons without thinking about it :)

**Comment by tstumm** [comment]
Fixed it for both favicons and the tray setting.

**Comment by tstumm** [comment]
The electron docs don't mention any side effects of hiding the BrowserWindow - although it should interrupt all repainting stuff. I guess (!) that it won't break anything, as the engine is still running, the window is just not shown. I didn't experience any issues, but this needs further investigation.

**Comment by kspearrin** [comment]
I can do some testing on macOS.

**Comment by kspearrin** [comment]
@tstumm I did some testing on macOS and made some changes to support the proper native icon there.

I tested on Ubuntu as well and it seems like no tray icon appears. The window disappears and I find no way to open it back without killing the app and starting it back up again. Any ideas about Linux tray icons?

**Comment by tstumm** [comment]
I read that `export XDG_CURRENT_DESKTOP=Unity` before running an electron app may fix this. It is related to Unity and I think they won't fix it, as Unity is no longer developed.

**Comment by kspearrin** [comment]
Maybe we should just support windows and mac for this feature.

**Comment by tstumm** [comment]
I'm not sure. Do you have any information how many Bitwarden users use Linux (for the desktop client)?
We could add a warning for Unity users though. But that does not seem to be a nice solution. I'll keep looking for other fixes.

**Comment by tstumm** [comment]
Does `sudo apt-get install libappindicator1` help? At least that's what https://electronjs.org/docs/api/tray says.

**Comment by tstumm** [comment]
And if I understood correctly, on Linux we need to listen for doubleclicks instead of single clicks.

**Comment by kspearrin** [comment]
After install libappindicator it now shows up in the tray.

image


Though single-clicking it just opens this context menu with the app name. Clicking the option in the context menu crashes the app.

**Comment by kspearrin** [comment]
Added a context menu for it here: https://github.com/bitwarden/desktop/commit/0e82da9bdc02048e61fb09c4c876f1005bd723d7

This seems to work now, but the behavior is still a bit weird. Opening the hidden window doesnt actually open it all the way back up. It ends up just reappearing in the task bar and then you still have to click it again to open the window.

image

**Comment by kspearrin** [comment]
Yes, it is also being used in the directory connector desktop application now.

**Pull request merged**